### PR TITLE
FindErrantGTIDs: superset is not an errant GTID situation

### PR DESCRIPTION
**Pull request by deepthi**
## Description
If there are only two replicas, and one has a GTIDSet that is a superset of the other, do not consider it as errant.
Of the two new unit tests, the second one (superset) fails without the code change in this PR.

This can cause serious issues, so I'm marking this for backport to all supported releases.

## Related Issue(s)
#16724 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by deepthi** [comment]
We'll need to follow up with an analysis of `FindErrantGTIDs` to see if there are other corner cases. Test coverage for that function is currently not good enough.
Also, as far back as https://github.com/vitessio/vitess/pull/6296#pullrequestreview-428655574 there were concerns about correctness, but they were not properly addressed when `FindErrantGTIDs` was implemented.
cc @shlomi-noach @GuptaManan100

**Review by frouioui — APPROVED** [review]

File: go/vt/vtctl/reparentutil/replication_test.go [hunk 1]
```diff
@@ -161,15 +161,15 @@ func TestFindValidEmergencyReparentCandidates(t *testing.T) {
...
+			name: "tablet with superset GTIDs is included",
 			statusMap: map[string]*replicationdatapb.StopReplicationStatus{
 				"r1": {
 					After: &replicationdatapb.Status{
 						SourceUuid:       "3E11FA47-71CA-11E1-9E33-C80AA9429562",
 						RelayLogPosition: "MySQL56/3E11FA47-71CA-11E1-9E33-C80AA9429562:1-5",
 					},
 				},
-				"errant": {
+				"r2": {
```
Comment (mattlord) [inline go/vt/vtctl/reparentutil/replication_test.go:172]: It's not clear to me how r2 does not have an errant GTID in this case. Isn't the test comparing the replica positions against the primary? r2 does in fact have a GTID that the primary does not have. Maybe I'm misunderstanding the test here though... should we be passing in a nil/empty primaryStatusMap for the ERS case?

Comment (deepthi) [reply go/vt/vtctl/reparentutil/replication_test.go:172]: primaryStatusMap has no effect on how we compute valid candidates. Its contents are simply stuffed into the positionMap if it is non-empty. The way the test is written does make it confusing.

Comment (deepthi) [reply go/vt/vtctl/reparentutil/replication_test.go:172]: I'll remove it from these tests so that it is less confusing. One of the things we need to revisit is whether when primaryStatusMap is non-empty we should be using that to determine errant GTIDs.

File: go/vt/vtctl/reparentutil/replication_test.go [hunk 2]
```diff
@@ -161,27 +161,41 @@ func TestFindValidEmergencyReparentCandidates(t *testing.T) {
 			shouldErr: false,
 		},
 		{
-			name: "tablet with errant GTIDs is excluded",
+			name: "tablet with superset GTIDs is included",
```
Comment (mattlord) [inline go/vt/vtctl/reparentutil/replication_test.go:164]: Nit, but I think the intention would be more clear with something like:
```suggestion
			name: "exactly 2 tablets, 1 with GTID superset, neither considered errant",
```

If it's clear to you though then no need.

**Review by mattlord — APPROVED** [review]
LGTM!

File: go/vt/vtctl/reparentutil/replication_test.go (hunk 2 above)
Comment (deepthi) [reply go/vt/vtctl/reparentutil/replication_test.go:164]: We can rewrite these tests altogether when we review FindErrantGTIDs with a larger group. I'm not satisfied with them.

**Comment by shlomi-noach** [comment]
The fix looks generally correct, and I'll need to dig into the `vtctl` logic that calls it to understand better. In the meantime, I do wonder why limit the fix to a 2-replica scenario? Why does the same logic not apply for a 3-replica setup?

**Comment by deepthi** [comment]
> The fix looks generally correct, and I'll need to dig into the `vtctl` logic that calls it to understand better. In the meantime, I do wonder why limit the fix to a 2-replica scenario? Why does the same logic not apply for a 3-replica setup?

That's a great question. Florent asked the same. I was trying to do the smallest fix for the known issue. We can extend the logic to >2 replicas in a follow up PR.
